src/encoding.rs

```diff
@@ -728,14 +766,11 @@ pub mod string {
     where
         B: Buf,
     {
+        let mut guard = StringGuard::new(value);
         unsafe {
-            // String::as_mut_vec is unsafe because it doesn't check that the bytes
-            // inserted into it the resulting vec are valid UTF-8. We check
-            // explicitly in order to ensure this is safe.
-            super::bytes::merge(wire_type, value.as_mut_vec(), buf)?;
-            str::from_utf8(value.as_bytes())
-                .map_err(|_| DecodeError::new("invalid string value: data is not UTF-8 encoded"))?;
+            super::bytes::merge(wire_type, guard.as_mut_vec(), buf)?;
```

Comment:
I think using some clever buffer swaps we can simplify this and remove the need for `unsafe` entirely, here's what I'm thinking:

```
    pub fn merge<B>(wire_type: WireType, value: &mut String, buf: &mut B) -> Result<(), DecodeError>
    where
        B: Buf,
    {
        let mut value_bytes = mem::replace(value, String::new()).into_bytes();
        bytes::merge(wire_type, &mut value_bytes, buf)?;
        *value = String::from_utf8(value_bytes)
            .map_err(|_| DecodeError::new("invalid string value: data is not UTF-8 encoded"))?;
        Ok(())
    }
```

That can even be simplified a bit more if we decide we need to overwrite on merge instead of append.  WDYT?

Reply:
@danburkert I think your implementation is good, unless somebody cares about the existing string data being lost on merge failure.